### PR TITLE
Feat: discussion topic and speed grader

### PR DESCRIPTION
**Pull request by IlyaRazuvaev**
https://app.clickup.com/t/86c04w6ay

### CODE REVIEW

File: src/discussions/speed-grader.ts [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        );
+    }
+
+    private iframeHandler(
+        iframe: HTMLIFrameElement,
+        key: number,
+        courseNumber: string,
+        topicNumber: string
+    ): void {
+        if (!iframe.src.includes('external_tools')) {
```
Comment (genadis) [inline src/discussions/speed-grader.ts]: what if it's not Annoto?

File: src/main.ts [hunk 2]
```diff
@@ -31,14 +25,20 @@ class AnnotoCanvas {
...
 
     setup(): void {
+        const discussionTopicHandler = new DiscussionTopicHandler(log);
+        const speedGraderHandler = new SpeedGraderHandler(log);
         if (this.isSetup) {
             log.warn('AnnotoCanvas: already setup');
             return;
         }
+
+        window.addEventListener('load', () => {
```
Comment (genadis) [inline src/main.ts]: what if page already loaded when setup is called?

File: src/discussions/speed-grader.ts (hunk 1 above)
Comment (genadis) [reply src/discussions/speed-grader.ts]: why not implement filter at iframes.forEach().filter

File: src/discussions/speed-grader.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+const USER_CONTENT_LOADED_INTERVAL = 200;
+const MAX_ATTEMPTS = 900;
+
+export class SpeedGraderHandler {
+    private studentId: string = '';
+    private observer: MutationObserver | undefined;
+    private threadInitSubscriptionDone: Record<string, boolean> = {};
+
+    constructor(private log: Pick<Console, 'log' | 'info' | 'warn' | 'error'>) {
```
Comment (genadis) [inline src/discussions/speed-grader.ts]: if it's reused, implement interface/type for the log

File: src/discussions/discussion-topic.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        if (this.label) {
+            this.log.info(`Discussion label: ${this.label}`);
+        } else {
+            this.log.info('Discussion label not found');
+        }
+    }
+
+    private handleDiscussionHolderMutations(): void {
+        const iframes = document.querySelectorAll('iframe');
+        iframes.forEach((iframe, key) => this.iframeHandler(iframe, key));
```
Comment (genadis) [inline src/discussions/discussion-topic.ts]: filter out if not annoto

**Review by genadis — CHANGES_REQUESTED** [review]

File: src/discussions/speed-grader.ts (hunk 1 above)
Comment (IlyaRazuvaev) [reply src/discussions/speed-grader.ts]: It is iterator, but you are right, it will be more reasonable to filter before call

Comment (IlyaRazuvaev) [reply src/discussions/speed-grader.ts]: Unfortunately I didn't find any option to understand that it is Annoto:
image
It was a possible solution to limit attempt amount to iframe, but canvas doesn't load internals of iframe until scroll and I couldn't understand is it another external tool, or it just not loaded
Example with not loaded annoto iframe
image

Comment (IlyaRazuvaev) [reply src/discussions/speed-grader.ts]: @IlyaRazuvaev Check cokie and UUID

**Review by genadis — APPROVED** [review]

**Pull request merged**